runtime/json-functions.cpp

```diff
@@ -0,0 +1,539 @@
+                  } else if (num < 0xffff) {
+                    value[l++] = static_cast<char>(0xe0 + (num >> 12));
+                    value[l++] = static_cast<char>(0x80 + ((num >> 6) & 63));
+                    value[l] = static_cast<char>(0x80 + (num & 63));
+                  } else {
+                    value[l++] = static_cast<char>(0xf0 + (num >> 18));
+                    value[l++] = static_cast<char>(0x80 + ((num >> 12) & 63));
+                    value[l++] = static_cast<char>(0x80 + ((num >> 6) & 63));
+                    value[l] = static_cast<char>(0x80 + (num & 63));
+                  }
+                  break;
+                }
+                /* fallthrough */
+              default:
+                return false;
+            }
+            i++;
+          } else {
+            value[l] = s[i++];
+          }
+        }
+        value.shrink(l);
+
+        new(&v) mixed(value);
+        i++;
+        return true;
+      }
+      break;
+    }
+    case '[': {
+      array<mixed> res;
+      i++;
+      json_skip_blanks(s, i);
+      if (s[i] != ']') {
+        do {
+          mixed value;
+          if (!do_json_decode(s, s_len, i, value)) {
+            return false;
+          }
+          res.push_back(value);
+          json_skip_blanks(s, i);
+        } while (s[i++] == ',');
+
+        if (s[i - 1] != ']') {
+          return false;
+        }
+      } else {
+        i++;
+      }
+
+      new(&v) mixed(res);
+      return true;
+    }
+    case '{': {
+      array<mixed> res;
+      i++;
+      json_skip_blanks(s, i);
+      if (s[i] != '}') {
+        do {
+          mixed key;
+          if (!do_json_decode(s, s_len, i, key) || !key.is_string()) {
+            return false;
+          }
+          json_skip_blanks(s, i);
+          if (s[i++] != ':') {
+            return false;
+          }
+
+          if (!do_json_decode(s, s_len, i, res[key])) {
+            return false;
+          }
+          json_skip_blanks(s, i);
+        } while (s[i++] == ',');
+
+        if (s[i - 1] != '}') {
+          return false;
+        }
+      } else {
+        i++;
+      }
+
+      new(&v) mixed(res);
+      return true;
+    }
+    default: {
+      int j = i;
+      while (s[j] == '-' || ('0' <= s[j] && s[j] <= '9') || s[j] == 'e' || s[j] == 'E' || s[j] == '+' || s[j] == '.') {
+        j++;
+      }
+      if (j > i) {
+        int64_t intval = 0;
+        if (php_try_to_int(s + i, j - i, &intval)) {
+          i = j;
+          new(&v) mixed(intval);
+          return true;
+        }
+
+        char *end_ptr;
+        double floatval = strtod(s + i, &end_ptr);
+        if (end_ptr == s + j) {
+          i = j;
+          new(&v) mixed(floatval);
+          return true;
+        }
+      }
+      break;
+    }
+  }
+
+  return false;
+}
+
+} // namespace
+
+mixed f$json_decode(const string &v, bool assoc) noexcept {
+  static_cast<void>(assoc);
```

Comment:
In php, when assoc = true, an associative array is returned (that's KPHP case). And when assoc = false, stdClass is returned.
We just skip this argument and work as if it's true always.
Maybe, if it's passed false, invoke php_warning?

Reply:
It was warning before, and it had been commented, I don't know why
![image](https://user-images.githubusercontent.com/2287155/105853229-84d3d200-5ff6-11eb-87e2-f4e17b6d8b89.png)

Reply:
Seems strange. Hopefully, there aren't many warnings on a working site (since PHP behaves completely differently). I'd suggest to uncomment it, what do you think?

Reply:
Added comment about it.

Reply:
False is default value here, I think it may produce a huge number of warnings.